### PR TITLE
Push0 rematerializer

### LINKED ISSUE #14225: Adjust Yul optimizer to rematerialize zero literals.

### PR DESCRIPTION
**Pull request by nikola-matic**
Fixes https://github.com/ethereum/solidity/issues/14225

This one adjusts the cost model used by the rematerializer; for zero literals, cost is decremented in order to rematerialize zero literals, i.e. to transform
```
let _1 := 0
mstore(0x40, _1)
```
into 
```
let _1 := 0
mstore(0x40, 0)
```
Obviously, the `let _1 := 0` in the zero rematerialized code will later be removed by the unused pruner (`u`).

### CODE REVIEW

File: libyul/optimiser/Metrics.cpp [hunk 1]
```diff
@@ -158,6 +158,10 @@ void CodeCost::operator()(Literal const& _literal)
 	case LiteralKind::Number:
 		for (u256 n = u256(_literal.value.str()); n >= 0x100; n >>= 8)
 			cost++;
+		if (valueOfLiteral(_literal) == 0)
+			if (auto&& evmDialect = dynamic_cast<EVMDialect const*>(&m_dialect))
```
Comment (ekpyron) [inline libyul/optimiser/Metrics.cpp]: Not that I'm into too much nitpicking, but ``auto&&`` for something that's clearly a plain pointer strikes me as weird enough to comment :-). - I'd make it ``auto`` or ``auto const*``.

Comment (nikola-matic) [reply libyul/optimiser/Metrics.cpp]: `auto` it is.

**Comment by ekpyron** [comment]
For cases like this it helps to post the output of ``scripts/gas_diff_stats.py`` in a comment to get a summary of the gas changes (even though they already do look quite good scrolling through them)

**Comment by nikola-matic** [comment]
<details><summary>Click for a table of gas differences</summary>


| File name                                                                                    |   IR-optimized (%) |   Legacy-Optimized (%) |   Legacy (%) |
|----------------------------------------------------------------------------------------------|--------------------|------------------------|--------------|
| byte_array_to_storage_cleanup.sol                                                            |       -0.828014    |            0.044899    |            0 |
| constructor_with_params_diamond_inheritance.sol                                              |       -1.12431     |            0           |            0 |
| constructor_inheritance_init_order.sol                                                       |       -1.07164     |            0           |            0 |
| constructor_with_params_inheritance.sol                                                      |       -1.22611     |            0           |            0 |
| constructor_with_params.sol                                                                  |       -1.46936     |            0           |            0 |
| constructor_inheritance_init_order_2.sol                                                     |       -1.07164     |            0           |            0 |
| structs/memory_structs_nested_load.sol                                                       |       -0.00815764  |            0           |            0 |
| structs/copy_struct_array_from_storage.sol                                                   |       -0.047073    |            0           |            0 |
| structs/struct_containing_bytes_copy_and_delete.sol                                          |       -0.0104794   |           -0.000746965 |            0 |
| structs/structs.sol                                                                          |       -0.0133894   |            0           |            0 |
| structs/copy_substructures_from_mapping.sol                                                  |       -0.0115133   |            0.000821254 |            0 |
| structs/copy_to_mapping.sol                                                                  |       -0.0108012   |            0           |            0 |
| structs/struct_delete_storage_with_array.sol                                                 |       -0.0198416   |            0           |            0 |
| structs/copy_substructures_to_mapping.sol                                                    |       -0.00611339  |            0.000269224 |            0 |
| structs/struct_memory_to_storage_function_ptr.sol                                            |       -0.00543537  |            0           |            0 |
| structs/struct_copy_via_local.sol                                                            |       -0.00638028  |            0           |            0 |
| structs/struct_copy.sol                                                                      |       -0.00612721  |            0           |            0 |
| structs/copy_from_mapping.sol                                                                |       -0.00823066  |            0.000821301 |            0 |
| structs/conversion/recursive_storage_memory.sol                                              |       -0.00516297  |            0           |            0 |
| structs/calldata/calldata_struct_with_nested_array_to_storage.sol                            |        0.00690237  |            0           |            0 |
| viaYul/copy_struct_invalid_ir_bug.sol                                                        |       -0.0265463   |            0           |            0 |
| various/create_calldata.sol                                                                  |       -1.20996     |           -0.0428356   |            0 |
| various/negative_stack_height.sol                                                            |        0           |           -0.560279    |            0 |
| various/skip_dynamic_types_for_structs.sol                                                   |       -0.0308051   |           -0.000557491 |            0 |
| various/senders_balance.sol                                                                  |       -0.382579    |            0           |            0 |
| various/value_insane.sol                                                                     |       -0.739663    |            0           |            0 |
| various/address_code.sol                                                                     |       -1.30863     |            0.0280293   |            0 |
| various/swap_in_storage_overwrite.sol                                                        |       -0.00546996  |            0           |            0 |
| various/value_complex.sol                                                                    |       -0.743328    |            0           |            0 |
| various/destructuring_assignment.sol                                                         |       -0.00868296  |           -0.000412506 |            0 |
| various/erc20.sol                                                                            |       -0.621405    |            0           |            0 |
| various/staticcall_for_view_and_pure.sol                                                     |       -7.61896e-06 |            0           |            0 |
| constructor/no_callvalue_check.sol                                                           |       -0.0133939   |            0           |            0 |
| constructor/bytes_in_constructors_packer.sol                                                 |       -1.16362     |            0.00171377  |            0 |
| constructor/constructor_static_array_argument.sol                                            |       -0.752115    |            0           |            0 |
| constructor/constructor_arguments_external.sol                                               |       -1.42768     |            0           |            0 |
| constructor/arrays_in_constructors.sol                                                       |       -1.08328     |            0           |            0 |
| constructor/bytes_in_constructors_unpacker.sol                                               |       -0.480996    |            0.0143216   |            0 |
| externalContracts/ramanujan_pi.sol                                                           |       -4.22532     |            0           |            0 |
| externalContracts/prbmath_signed.sol                                                         |       -1.0965      |            0           |            0 |
| externalContracts/snark.sol                                                                  |       -0.0101587   |            0           |            0 |
| externalContracts/strings.sol                                                                |       -0.142526    |            0           |            0 |
| externalContracts/prbmath_unsigned.sol                                                       |       -0.487211    |            0           |            0 |
| externalContracts/base64.sol                                                                 |       -0.049224    |            0           |            0 |
| externalContracts/deposit_contract.sol                                                       |       -1.34765     |            0           |            0 |
| externalContracts/FixedFeeRegistrar.sol                                                      |       -6.98692     |            0           |            0 |
| immutable/use_scratch.sol                                                                    |       -1.03815     |            0           |            0 |
| immutable/multi_creation.sol                                                                 |       -0.677619    |            0           |            0 |
| functionCall/gas_and_value_basic.sol                                                         |       -2.34645     |            0           |            0 |
| functionCall/mapping_array_internal_argument.sol                                             |       -0.0358205   |            0           |            0 |
| functionCall/creation_function_call_with_args.sol                                            |       -0.222682    |            0           |            0 |
| functionCall/gas_and_value_brace_syntax.sol                                                  |       -2.34645     |            0           |            0 |
| functionCall/creation_function_call_with_salt.sol                                            |       -0.222495    |            0           |            0 |
| functionCall/failed_create.sol                                                               |       -2.94095     |            0           |            0 |
| abiEncoderV1/abi_encode_calldata_slice.sol                                                   |       -0.108917    |            0           |            0 |
| abiEncoderV1/abi_decode_v2_storage.sol                                                       |       -0.00836754  |            0.00246122  |            0 |
| abiEncoderV1/struct/struct_storage_ptr.sol                                                   |       -0.0234226   |            0           |            0 |
| salted_create/salted_create_with_value.sol                                                   |       -0.473179    |            0           |            0 |
| salted_create/salted_create.sol                                                              |       -2.03172e-06 |            0           |            0 |
| state/blockhash_basic.sol                                                                    |       -1.57265     |            0           |            0 |
| abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol                          |       -7.08694     |            0           |            0 |
| abiEncoderV2/storage_array_encoding.sol                                                      |       -0.0167075   |            0           |            0 |
| abiEncoderV2/calldata_array.sol                                                              |       -0.128334    |            0           |            0 |
| abiEncoderV2/abi_encode_calldata_slice.sol                                                   |       -0.108917    |            0           |            0 |
| abiEncoderV2/abi_encode_v2.sol                                                               |       -0.00976554  |            0           |            0 |
| abiEncoderV2/calldata_overlapped_dynamic_arrays.sol                                          |       -0.00985222  |            0           |            0 |
| userDefinedValueType/erc20.sol                                                               |       -2.083       |            0           |            0 |
| userDefinedValueType/calldata.sol                                                            |       -0.101151    |            0           |            0 |
| libraries/internal_types_in_library.sol                                                      |       -0.0837722   |            0           |            0 |
| libraries/using_library_mappings_return.sol                                                  |       -0.0540473   |            0           |            0 |
| libraries/using_library_mappings_public.sol                                                  |       -0.0100289   |            0           |            0 |
| abiencodedecode/abi_decode_simple_storage.sol                                                |       -0.0228457   |            0.00367531  |            0 |
| events/event_dynamic_nested_array_storage_v2.sol                                             |       -0.0334963   |           -0.00487715  |            0 |
| events/event_indexed_string.sol                                                              |       -0.372414    |            0           |            0 |
| events/event_dynamic_array_storage_v2.sol                                                    |       -0.0220245   |            0           |            0 |
| events/event_dynamic_array_storage.sol                                                       |       -0.0220245   |            0           |            0 |
| array/create_memory_array.sol                                                                |       -0.00635794  |            0           |            0 |
| array/array_storage_index_zeroed_test.sol                                                    |       -0.0824872   |            0           |            0 |
| array/array_storage_push_empty.sol                                                           |       -0.514734    |            0           |            0 |
| array/array_storage_index_boundary_test.sol                                                  |       -1.75151     |            0           |            0 |
| array/arrays_complex_from_and_to_storage.sol                                                 |        0.00911347  |            0           |            0 |
| array/function_array_cross_calls.sol                                                         |       -1.84787     |            0           |            0 |
| array/reusing_memory.sol                                                                     |       -0.211303    |            0           |            0 |
| array/invalid_encoding_for_storage_byte_array.sol                                            |       -0.0156865   |            0           |            0 |
| array/array_storage_push_empty_length_address.sol                                            |       -2.08801     |            0           |            0 |
| array/dynamic_arrays_in_storage.sol                                                          |       -0.452826    |            0           |            0 |
| array/array_storage_push_pop.sol                                                             |       -0.637499    |            0           |            0 |
| array/fixed_arrays_in_constructors.sol                                                       |       -1.0152      |            0           |            0 |
| array/bytes_length_member.sol                                                                |       -0.00724638  |           -0.000904323 |            0 |
| array/dynamic_multi_array_cleanup.sol                                                        |       -0.14156     |            0           |            0 |
| array/fixed_arrays_as_return_type.sol                                                        |       -1.40208     |            0           |            0 |
| array/dynamic_array_cleanup.sol                                                              |       -0.00192576  |            0           |            0 |
| array/array_storage_index_access.sol                                                         |       -0.14177     |            0           |            0 |
| array/constant_var_as_array_length.sol                                                       |       -0.919462    |            0           |            0 |
| array/array_storage_length_access.sol                                                        |       -0.944417    |            0           |            0 |
| array/fixed_array_cleanup.sol                                                                |       -0.00214904  |            0           |            0 |
| array/array_memory_index_access.sol                                                          |       -0.00962443  |            0           |            0 |
| array/byte_array_transitional_2.sol                                                          |       -0.296216    |            0           |            0 |
| array/pop/byte_array_pop_long_storage_empty.sol                                              |       -0.191698    |            0           |            0 |
| array/pop/array_pop_array_transition.sol                                                     |       -0.0240131   |            0           |            0 |
| array/pop/byte_array_pop_masking_long.sol                                                    |       -0.0777396   |            0           |            0 |
| array/pop/array_pop_uint16_transition.sol                                                    |       -0.0546677   |            0           |            0 |
| array/pop/byte_array_pop_long_storage_empty_garbage_ref.sol                                  |       -0.0386555   |            0           |            0 |
| array/pop/array_pop_uint24_transition.sol                                                    |        0.110376    |            0           |            0 |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol                       |       -0.0208849   |            0           |            0 |
| array/copying/memory_dyn_2d_bytes_to_storage.sol                                             |        0.122545    |           -0.00232502  |            0 |
| array/copying/array_copy_target_simple_2.sol                                                 |       -0.00600124  |            0           |            0 |
| array/copying/array_of_struct_memory_to_storage.sol                                          |       -0.0278733   |            0           |            0 |
| array/copying/array_copy_storage_storage_different_base_nested.sol                           |       -0.00943385  |            0           |            0 |
| array/copying/bytes_inside_mappings.sol                                                      |       -0.0111073   |            0           |            0 |
| array/copying/copy_byte_array_in_struct_to_storage.sol                                       |       -0.0108227   |           -0.000348492 |            0 |
| array/copying/array_copy_target_leftover.sol                                                 |       -0.125161    |            0           |            0 |
| array/copying/array_of_structs_containing_arrays_calldata_to_storage.sol                     |       -0.114718    |            0           |            0 |
| array/copying/cleanup_during_multi_element_per_slot_copy.sol                                 |       -3.44255     |            0           |            0 |
| array/copying/array_copy_storage_storage_different_base.sol                                  |       -0.00829018  |            0           |            0 |
| array/copying/copying_bytes_multiassign.sol                                                  |        0.00992601  |           -0.000900325 |            0 |
| array/copying/array_storage_multi_items_per_slot.sol                                         |       -0.0128854   |            0           |            0 |
| array/copying/nested_array_of_structs_storage_to_storage.sol                                 |       -0.01949     |            0           |            0 |
| array/copying/array_copy_storage_storage_struct.sol                                          |       -0.0922094   |            0           |            0 |
| array/copying/array_to_mapping.sol                                                           |       -0.0162281   |           -0.00476489  |            0 |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol                                 |       -0.0225309   |            0           |            0 |
| array/copying/array_elements_to_mapping.sol                                                  |       -0.0379846   |           -0.00470794  |            0 |
| array/copying/nested_array_of_structs_memory_to_storage.sol                                  |        0.0229701   |            0           |            0 |
| array/copying/elements_of_nested_array_of_structs_memory_to_storage.sol                      |        0.0444671   |            0           |            0 |
| array/copying/array_copy_including_array.sol                                                 |       -0.00205021  |            0           |            0 |
| array/copying/array_nested_memory_to_storage.sol                                             |       -0.0127741   |            0           |            0 |
| array/copying/storage_memory_nested_from_pointer.sol                                         |       -0.0451513   |            0           |            0 |
| array/copying/bytes_storage_to_storage.sol                                                   |        0.149621    |           -0.00789889  |            0 |
| array/copying/array_copy_storage_to_memory_nested.sol                                        |       -0.0544167   |           -0.00437661  |            0 |
| array/copying/copy_removes_bytes_data.sol                                                    |       -0.00451002  |           -0.000563209 |            0 |
| array/copying/calldata_array_dynamic_to_storage.sol                                          |       -0.00720916  |            0           |            0 |
| array/copying/copy_function_internal_storage_array.sol                                       |       -0.018779    |            0           |            0 |
| array/copying/nested_array_element_storage_to_storage.sol                                    |       -0.0162479   |            0           |            0 |
| array/copying/array_of_function_external_storage_to_storage_dynamic_different_mutability.sol |       -0.0134255   |            0           |            0 |
| array/copying/calldata_array_to_mapping.sol                                                  |       -0.15725     |            0           |            0 |
| array/copying/array_of_struct_calldata_to_storage.sol                                        |       -0.0375827   |            0           |            0 |
| array/copying/array_copy_storage_storage_static_dynamic.sol                                  |       -0.00487242  |            0           |            0 |
| array/copying/elements_of_nested_array_of_structs_calldata_to_storage.sol                    |        0.0103475   |            0           |            0 |
| array/copying/array_of_function_external_storage_to_storage_dynamic.sol                      |       -0.0133836   |            0           |            0 |
| array/copying/array_nested_calldata_to_storage.sol                                           |       -0.0148914   |            0           |            0 |
| array/copying/array_copy_nested_array.sol                                                    |       -0.00695941  |            0           |            0 |
| array/copying/array_copy_different_packing.sol                                               |       -0.037478    |            0           |            0 |
| array/copying/array_copy_target_simple.sol                                                   |       -0.00438635  |            0           |            0 |
| array/copying/array_copy_cleanup_uint40.sol                                                  |       -0.0920175   |            0           |            0 |
| array/copying/nested_array_of_structs_calldata_to_storage.sol                                |        0.0121485   |            0           |            0 |
| array/copying/array_copy_storage_storage_dyn_dyn.sol                                         |       -0.0161629   |            0           |            0 |
| array/copying/storage_memory_nested_struct.sol                                               |       -0.0370936   |            0           |            0 |
| array/copying/array_copy_calldata_storage.sol                                                |       -0.000616976 |            0           |            0 |
| array/copying/copy_byte_array_to_storage.sol                                                 |       -0.00976113  |           -0.00136312  |            0 |
| array/copying/storage_memory_packed_dyn.sol                                                  |       -0.0965356   |            0           |            0 |
| array/copying/storage_memory_nested.sol                                                      |       -0.0451513   |            0           |            0 |
| array/copying/array_copy_storage_storage_static_static.sol                                   |       -0.00968015  |            0           |            0 |
| array/copying/array_copy_clear_storage.sol                                                   |       -0.0325618   |            0           |            0 |
| array/copying/nested_dynamic_array_element_calldata_to_storage.sol                           |        0.000661625 |            0           |            0 |
| array/copying/function_type_array_to_storage.sol                                             |       -0.00600147  |            0           |            0 |
| array/copying/storage_memory_nested_bytes.sol                                                |       -0.0330367   |           -0.00540891  |            0 |
| array/delete/bytes_delete_element.sol                                                        |       -0.114384    |            0.00202522  |            0 |
| array/push/push_no_args_2d.sol                                                               |       -0.0207792   |            0           |            0 |
| array/push/array_push_nested_from_calldata.sol                                               |       -0.0106359   |            0           |            0 |
| array/push/array_push.sol                                                                    |       -0.034111    |            0           |            0 |
| array/push/nested_bytes_push.sol                                                             |       -0.030698    |            0.00498855  |            0 |
| array/push/array_push_struct_from_calldata.sol                                               |       -0.04085     |            0           |            0 |
| array/push/push_no_args_bytes.sol                                                            |        0.0740286   |            0           |            0 |
| array/push/byte_array_push_transition.sol                                                    |       -0.657424    |            0           |            0 |
| array/push/array_push_struct.sol                                                             |       -0.0865391   |            0           |            0 |
| inheritance/value_for_constructor.sol                                                        |       -1.71084     |            0           |            0 |
| smoke/constructor.sol                                                                        |       -1.84444     |            0           |            0 |
| isoltestTesting/balance_other_contract.sol                                                   |       -1.13461     |            0           |            0 |
| storage/packed_storage_structs_bytes.sol                                                     |       -0.00905626  |            0           |            0 |
| storage/empty_nonempty_empty.sol                                                             |       -0.0151928   |            0.00445991  |            0 |
| calldata/copy_from_calldata_removes_bytes_data.sol                                           |       -0.0154684   |           -0.000643919 |            0 |

</details>

File: test/libyul/yulOptimizerTests/fullSuite/unusedFunctionParameterPruner_loop.yul [hunk 2]
```diff
@@ -25,13 +27,12 @@
...
-//         let _1 := 0
-//         let a := calldataload(_1)
-//         let _2 := iszero(a)
+//         let a := calldataload(0)
+//         let _1 := iszero(a)
 //         for { } iszero(b) { b := add(b, not(0)) }
 //         {
-//             mstore(a, _1)
-//             if _2 { leave }
+//             mstore(a, 0)
```
Comment (ekpyron) [inline test/libyul/yulOptimizerTests/fullSuite/unusedFunctionParameterPruner_loop.yul:34]: Actually, these tests indicate that this change already does the trick in general and that the CSE in the cleanup sequence does not destroy this again - probably due to the fact that we don't run the expression splitter again beforehand. So maybe adding ``mu`` to the cleanup sequence is even unnecessary after this (we should still check and evaluate, though).

File: Changelog.md [hunk 3]
```diff
@@ -4,6 +4,7 @@ Language Features:
 
 
 Compiler Features:
+ * Yul Optimizer: Rematerialize zero literals in order to take advantage of the PUSH0 opcode (introduced in Shanghai).
```
Comment (ekpyron) [inline Changelog.md]: We could describe this in a more high-level manner, e.g. something like:

```suggestion
 * Yul Optimizer: If ``PUSH0`` is supported, favour zero literals over storing zero values in variables.
```

But the current entry is also fine, if it quotes ``PUSH0`` (I trust that you'll do that despite me already approving the rest ;-)).

**Review by ekpyron — DISMISSED** [review]

**Review by ekpyron — APPROVED** [review]

**Pull request merged**